### DISCUSSION TITLE
rbenv install 3.1.4 fails on Mac M4 15.0.1 (Sequoia)

### DISCUSSION THREAD

**Discussion by dryabinin94** [rbenv/ruby-build#2453] [Build failures]
> rbenv install 3.1.4

**Installation process log:**

> Downloading ruby-build...
> ruby-build: using openssl@3 from homebrew
> ==> Downloading ruby-3.1.4.tar.gz...
> -> curl -q -fL -o ruby-3.1.4.tar.gz https://cache.ruby-lang.org/pub/ruby/3.1/ruby-3.1.4.tar.gz
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
> 100 19.9M  100 19.9M    0     0  8734k      0  0:00:02  0:00:02 --:--:-- 8733k
> ==> Installing ruby-3.1.4...
> ruby-build: using readline from homebrew
> ruby-build: using libyaml from homebrew
> ruby-build: using gmp from homebrew
> -> ./configure "--prefix=$HOME/.asdf/installs/ruby/3.1.4" --with-openssl-dir=/usr/local/opt/openssl@3 --enable-shared --with-readline-dir=/usr/local/opt/readline --with-libyaml-dir=/usr/local/opt/libyaml --with-gmp-dir=/usr/local/opt/gmp --with-ext=openssl,psych,+
> -> make -j 12
> 
> BUILD FAILED (macOS 15.0.1 on arm64 using ruby-build 20241007)
> 
> You can inspect the build directory at /var/folders/x0/p185qzl11xl391xd6379sqz80000gn/T/ruby-build.20241016105503.76864.KctVWB
> See the full build log at /var/folders/x0/p185qzl11xl391xd6379sqz80000gn/T/ruby-build.20241016105503.76864.log

Full log - [ruby-build.20241016105503.76864.log](https://github.com/user-attachments/files/17391685/ruby-build.20241016105503.76864.log)

Interesting place in log with error:

`readline.c:1903:37: error: use of undeclared identifier 'username_completion_function'; did you mean 'rl_username_completion_function'?
 1903 |                                     rl_username_completion_function);
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                     rl_username_completion_function
readline.c:79:42: note: expanded from macro 'rl_username_completion_function'
   79 | # define rl_username_completion_function username_completion_function
      |                                          ^
/usr/local/opt/readline/include/readline/readline.h:494:14: note: 'rl_username_completion_function' declared here
  494 | extern char *rl_username_completion_function (const char *, int);`

**Comment by mislav** [comment]
This is an often-reported error, but I have yet to determine how to reliably reproduce it or solve it.

From what you've shared, I'm seeing that you have an M1 or M2 Apple machine, but that your Homebrew installation is in `/usr/local`. On Apple silicon architecture, Homebrew is usually placed in `/opt/homebrew`, and the fact that yours is in an old location makes me guess that you've either ported your setup from a previous machine running Intel CPU architecture, or that you're running Homebrew in emulated "x64" mode instead of native "arm64". This was a popular solution a while ago to ease transition to Apple silicon, but causes problems in the long run.

What you could try:

1. Reinstalling readline with `brew reinstall readline`;
2. Reinstalling/moving Homebrew entirely to `/opt/homebrew`. I don't know how this is easiest done; probably just by uninstalling old Homebrew and reinstalling it and all formula from scratch.

Before reinstalling readline, however, you could check `brew info readline` to get version information about your current version (the one that is causing problems) and post it here for posterity. If there is a specific version of readline that is not compatible with Ruby, I think it's good for everyone to know.

If the error is really persistent, then it would be helpful for you to share the copy of your `ext/readline/mkmf.log` log file, where debugging information regarding the compilation of the readline extension will be stored. This could be found under the temporary build location for ruby as reported by ruby-build.
